dali/pipeline/operators/expressions/arithmetic.h

```diff
@@ -181,15 +189,20 @@ DLL_PUBLIC kernels::TensorListShape<> PropagateShapes(ExprNode &expr,
     expr.SetShape(ws.template InputRef<Backend>(e.GetInputIndex()).shape());
     return expr.GetShape();
   }
-  if (expr.GetSubexpressionCount() == 2) {
-    auto &func = dynamic_cast<ExprFunc&>(expr);
-    auto left_shape = PropagateShapes<Backend>(func[0], ws);
-    auto right_shape = PropagateShapes<Backend>(func[1], ws);
-    func.SetShape(ShapePromotion(func.GetFuncName(), left_shape, right_shape));
-    return func.GetShape();
+  auto &func = dynamic_cast<ExprFunc &>(expr);
+  int subexpressions_count = expr.GetSubexpressionCount();
+  DALI_ENFORCE(subexpressions_count == 1 || subexpressions_count == 2,
+               "Only unary and binary expressions are supported");
+
+  // TODO(klecki): cannot have a SmallVector here as compiler complains that calling __host__
```

Comment:
GetShape returns a reference to a member. You can leverage this fact and have a small vector of pointers to shapes here instead of needlessly copying these relatively heavy objects.

Reply:
Done, now I'm forwarding the reference.